Crypto/ECC.hs

```diff
@@ -0,0 +1,236 @@
+        len = 48 -- (384 + 7) `div` 8
+        shared = SharedSecret $ i2ospOf_ len x
+
+data Curve_P521R1 = Curve_P521R1
+
+instance EllipticCurve Curve_P521R1 where
+    newtype Point Curve_P521R1 = P521Point { unP521Point :: H.Point } deriving (Eq,Show)
+    newtype Scalar Curve_P521R1 = P521Scalar { unP521Scalar :: H.PrivateNumber } deriving (Eq,Show)
+    curveGetOrder _ = H.ecc_n $ H.common_curve $ H.getCurveByName H.SEC_p521r1
+    curveGetBasePoint = P521Point $ H.ecc_g $ H.common_curve $ H.getCurveByName H.SEC_p521r1
+    curveOfScalar _ = Curve_P521R1
+    curveOfPoint _ = Curve_P521R1
+    curveGenerateScalar = P521Scalar <$> H.scalarGenerate (H.getCurveByName H.SEC_p521r1)
+    curveGenerateKeyPair = toKeyPair <$> H.scalarGenerate (H.getCurveByName H.SEC_p521r1)
+      where toKeyPair scalar = KeyPair (P521Point $ H.pointBaseMul (H.getCurveByName H.SEC_p521r1) scalar) (P521Scalar scalar)
+    encodePoint (P521Point (H.Point x y)) = encodeECPoint x y 66
+    encodePoint (P521Point _)             = error "encodePoint P521"
+    decodePoint bs = P521Point $ H.Point x y
+      where
+        (x,y) = decodeECPoint bs
+
+instance EllipticCurveArith Curve_P521R1 where
+    pointAdd a b = P521Point $ (H.pointAdd (H.getCurveByName H.SEC_p521r1) `on` unP521Point) a b
+    pointSmul s p = P521Point (H.pointMul (H.getCurveByName H.SEC_p521r1) (unP521Scalar s) (unP521Point p))
+
+instance EllipticCurveDH Curve_P521R1 where
+    ecdh s p = shared
+      where
+        H.Point x _ = unP521Point $ pointSmul s p
+        len = 66 -- (521 + 7) `div` 8
+        shared = SharedSecret $ i2ospOf_ len x
+
+data Curve_X25519 = Curve_X25519
+
+instance EllipticCurve Curve_X25519 where
+    newtype Point Curve_X25519 = X25519Point X25519.PublicKey deriving (Eq,Show)
+    newtype Scalar Curve_X25519 = X25519Scalar X25519.SecretKey deriving (Eq,Show)
+    curveGetOrder     _ = undefined
+    curveGetBasePoint = undefined
+    curveOfScalar _ = Curve_X25519
+    curveOfPoint _ = Curve_X25519
+    curveGenerateScalar = X25519Scalar <$> X25519.generateSecretKey
+    curveGenerateKeyPair = do
+        s <- X25519.generateSecretKey
+        let p = X25519.toPublic s
+        return $ KeyPair (X25519Point p) (X25519Scalar s)
+    encodePoint (X25519Point p) = X25519.fromPublicKey p
+    decodePoint bs = X25519Point $ X25519.toPublicKey bs
+
+instance EllipticCurveArith Curve_X25519 where
+    pointAdd  = undefined
+    pointSmul = undefined
+
+instance EllipticCurveDH Curve_X25519 where
+    ecdh (X25519Scalar s) (X25519Point p) = SharedSecret $ convert secret
+      where
+        secret = X25519.dh p s
+
+encodeECPoint :: Integer -> Integer -> Int -> ByteString
+encodeECPoint x y siz = B.concat [uncompressed,xb,yb]
+  where
+    uncompressed = B.singleton 4
+    xb = i2ospOf_ siz x
+    yb = i2ospOf_ siz y
+
+decodeECPoint :: ByteString -> (Integer,Integer)
+decodeECPoint mxy = (x,y)
+  where
+    xy = B.drop 1 mxy -- dropping 4 (uncompressed)
```

Comment:
it really really shouldn't be a drop 1. at the very least it should be a case that check for 4, and put a note for the 2,3 format. Also this shouldn't use `ByteString` but should do things through the ByteArray / ByteArrayClass